### PR TITLE
Endpoints for Tagging Functionality / Database Migrations and Seeding

### LINKED ISSUE #14: BE: Chore: Seed Tags
Starter seed list:

javascript
react
ruby
rails
video
blog
tutorial
python
java
C#
conference talk
Angular
Vue.js
Django
Flask
Spring
Laravel
Express.js
next.js
.NET
Bootstrap
jQuery
TensorFlow
PyTorch
Node.js
testing
rspec
minitest
cypress
mocha
chai
Documentation
Tutorials
Courses
Code Samples
Books
Podcasts
Webinars
Cheat Sheets
Articles
Whitepapers
Tools
Reference Guides
Workshops
Object-Oriented Programming (OOP)
Functional Programming
Procedural Programming
Test-Driven Development (TDD)
Agile
DevOps
Microservices
RESTful APIs
Background Jobs
GraphQL
Serverless
SQL
NoSQL
PostgreSQL
MySQL
MongoDB
Redis
Firebase
SQLite
Oracle
Git
GitHub
GitLab
Bitbucket
SVN
AWS
Google Cloud
Azure
Heroku
DigitalOcean
Kubernetes
Docker
Render
Vercel
Fly.io
HTML
CSS
JavaScript
SASS/SCSS
Webpack
Gulp
Babel
Responsive Design
GraphQL
Authentication
Authorization
Middleware
WebSockets
Encryption
SSL/TLS
OAuth
JWT
Secure Coding Practices
Penetration Testing
Vulnerability Scanning
Unit Testing
Integration Testing
End-to-End Testing
Test Automation
Load Testing
Selenium
Jest
Mocha
RSpec
Cypress
Agile
Scrum
Kanban
Jira
Trello
Asana
Time Tracking
Continuous Deployment
Continuous Integration
Jenkins
CircleCI
Travis CI
Docker
Kubernetes
Ansible
Terraform
Design Patterns
SOLID Principles
Architecture
Domain-Driven Design (DDD)
Model-View-Controller (MVC)
Event-Driven Architecture
Monolithic Architecture
Clean Architecture
Neural Networks
Deep Learning
Natural Language Processing (NLP)
Computer Vision
AI Ethics
Model Training
Data Preprocessing
Model Evaluation
iOS Development
Android Development
Flutter
React Native
SwiftUI
Xamarin
Kotlin Multiplatform
IDEs
Text Editors
Command-Line Tools
Debugging Tools
Package Managers
npm
pip
bundler
Profiling Tools
Code Linters
REST
SOAP
GraphQL
JSON
XML
OpenAPI
gRPC
WebSocket
User Experience (UX)
User Interface (UI)
Wireframing
Prototyping
Figma
Sketch
Adobe XD
User Research
Open Source
Licensing
Ethics in Software Development
Software Maintenance
Code Review
Documentation Writing

### LINKED ISSUE #16: BE: POST Add a tag to a link
Endpoint for adding a tag to a link. FE passes link and tag as params and it is stored in a LinkTags joins table

### LINKED ISSUE #17: BE: GET all tags for a given link
When a shortened link is passed, we return an index of all tags that have been associated with it

### LINKED ISSUE #18: BE: GET all tags from BE database
This endpoint will pull the full list of seeded tag names from the backend database

### LINKED ISSUE #20: BE: Chore: Database Migrations for Links and LinkTags joins table
create backend database migrations for tags and LinkTags joins table

### LINKED ISSUE #23: BE: DELETE remove tag from link
Removes an associated tag from a link

### PR DESCRIPTION
**Pull request by jdmchugh111**
## Description of Changes
Migrations for Tags table and LinkTags joins table. All Tags seeded in seeds.rb. All 4 tagging endpoints functional and fully tested

## What are the relevant tickets (if any) 

closes #20 
closes #14 
closes #18 
closes #16 
closes #23 
closes #17

## Testing
Unit testing for models. Integration testing for requests - all tests passing and coverage is 100%

## Checklist
- [X] The code follows the project's coding standards.
- [X] Unit tests have been added or updated to cover the changes.
- [X] Documentation has been updated to reflect the changes (if applicable).
- [X] The code compiles without errors.
- [X] The changes have been tested locally and pass all relevant tests.
- [X] All new and existing tests pass.
- [ ] The pull request has been reviewed by at least one other contributor.

### CODE REVIEW

**Comment by epintozzi** [comment]
1) This is a GIANT PR. Typically I'd prefer to see each action (index, show, create, update, destroy) as its own PR. It's really hard to review big PRs well. Smaller ones set us up for better success. 

2) I'm curious why you chose to put everything in a `TagsController` instead of a `LinkTagsController` for the CRUD functionality of adding tags to links?

**Comment by jdmchugh111** [comment]
On #1 - I definitely agree that it's very big for one PR. I just wanted to set up the FE team to be able to integrate the backend tagging endpoints asap. Would it have made more sense to separate endpoints into their own PRs and then open multiple PRs all at once?

On #2 - I was mostly going for ease of use / developer-friendly so I thought the api/v1/tags made most sense for the base path of the url, and Rails routes this to the TagsController by default. I can go in and set a custom controller so that they are handled in a LinkTagsController instead. I can also change the path for the endpoints if you think the current one doesn't feel conventional or should be something more like api/v1/link_tags instead

**Comment by noahdurbin** [comment]
I'll wait to merge this until your questions are answered James but I think this looks really good overall. everything made sense to me, I'm in a good place to get started on the next endpoints.

**Comment by epintozzi** [comment]
Noah if you’re good with this go ahead and merge. 

James- yeah ideally each endpoint would’ve been its own PR but it’s not worth the effort to do that this time around. Just something to keep in mind for future PRs. As for the other question (#2) that’s a design decision thing. It’s slightly unrestful but you’re allowed to make choices that break convention. I just need you to be able to talk about why. You have a reason for why you did it that way. No need to change.
